services/system/Transact/src/RTransact.cpp

```diff
@@ -177,129 +179,327 @@ namespace
 
-      for (auto client : row->clients)
-         if (client.json)
+      PSIBASE_SUBJECTIVE_TX
+      {
+         for (auto id : txids)
          {
-            psio::get<0>(action->rawData().value()) = client.socket;
-            call(action.data(), action.size());
+            pendingTxTable.erase(id);
+            dataTable.erase(id);
          }
+      }
    }
-   if (bin)
+
+   void sendReply(const psibase::Checksum256&                 id,
+                  psio::view<const psibase::TransactionTrace> trace,
+                  const ClientFilter&                         clientFilter = noFilter)
    {
-      FracpackHttpReply<TransactionTraceRef&> reply{.contentType = "application/octet-stream",
-                                                    .body{pruned}};
-      ActionViewBuilder<HttpServer>           http{getReceiver(), HttpServer::service};
-      auto                                    action = http.sendReply(0, reply);
+      TransactionTraceRef pruned     = PruneTrace{true}(trace);
+      auto [jsonClients, binClients] = claimClientReply(id, clientFilter);
+
+      if (jsonClients.empty() && binClients.empty())
+         return;
 
-      for (auto client : row->clients)
-         if (!client.json)
+      ActionViewBuilder<HttpServer> http{getReceiver(), HttpServer::service};
+
+      if (!jsonClients.empty())
+      {
+         JsonHttpReply<TransactionTraceRef&> reply{.contentType = "application/json",
+                                                   .body{pruned}};
+         auto                                action = http.sendReply(0, reply);
+
+         for (auto socket : jsonClients)
          {
-            psio::get<0>(action->rawData().value()) = client.socket;
+            psio::get<0>(action->rawData().value()) = socket;
             call(action.data(), action.size());
          }
+      }
+
+      if (!binClients.empty())
+      {
+         FracpackHttpReply<TransactionTraceRef&> reply{.contentType = "application/octet-stream",
+                                                       .body{pruned}};
+         auto                                    action = http.sendReply(0, reply);
+
+         for (auto socket : binClients)
+         {
+            psio::get<0>(action->rawData().value()) = socket;
+            call(action.data(), action.size());
+         }
+      }
    }
+
+   void sendReplies(const std::vector<psibase::Checksum256>& txids)
+   {
+      stopTracking(txids);
+      auto successfulTxs = RTransact::WriteOnly{}.open<TxSuccessTable>();
+      auto failedTxs     = RTransact::Subjective{}.open<TxFailedTable>();
+
+      for (auto id : txids)
+      {
+         std::optional<std::vector<char>> trace;
+
+         if (auto tx = successfulTxs.get(id))
+         {
+            trace = psio::to_frac(tx->trace);
+            successfulTxs.erase(id);
+         }
+
+         PSIBASE_SUBJECTIVE_TX
+         {
+            if (auto tx = failedTxs.get(id))
+            {
+               if (!trace)
```

Comment:
If you switch these conditions and make erase unconditional, it will avoid loading the failed trace into memory, when there's already a success trace.

Reply:
Also, if you already have an object from `get`, `remove` is better than `erase`.